kimchi/src/snarky/boolean.rs

```diff
@@ -0,0 +1,192 @@
+        if xs.is_empty() {
+            return Self::false_(); // TODO: shouldn't we panic instead?
+        } else if xs.len() == 1 {
+            return xs[0].clone();
+        } else if xs.len() == 2 {
+            return xs[0].or(&xs[1], cs); // TODO: is this better than below?
+        }
+
+        let zero = CVar::Constant(F::zero());
+
+        let xs: Vec<_> = xs.into_iter().map(|x| &x.0).collect();
+        let sum = CVar::sum(&xs);
+        let all_zero = sum.equal(cs, &zero);
+
+        all_zero.not()
+    }
+
+    pub fn all(xs: &[Self], cs: &mut RunState<F>) -> Self {
+        if xs.is_empty() {
+            return Self::true_(); // TODO: shouldn't we panic instead?
+        } else if xs.len() == 1 {
+            return xs[0].clone();
+        } else if xs.len() == 2 {
+            return xs[0].and(&xs[1], cs); // TODO: is this better than below?
+        }
+
+        let expected = CVar::Constant(F::from(xs.len() as u64));
+        let xs: Vec<_> = xs.iter().map(|x| &x.0).collect();
+        let sum = CVar::sum(&xs);
+
+        sum.equal(cs, &expected)
+    }
+
+    pub fn to_constant(&self) -> Option<bool> {
+        match self.0 {
+            CVar::Constant(x) => Some(x == F::one()),
+            _ => None,
+        }
+    }
+
+    pub fn xor(&self, other: &Self, state: &mut RunState<F>) -> Self {
+        match (self.to_constant(), other.to_constant()) {
+            (Some(x), Some(y)) => {
+                // (var_of_value (Caml.not (Bool.equal b1 b2)))
+                todo!()
+            }
+            (Some(true), None) => other.not(),
+            (None, Some(true)) => self.not(),
+            (Some(false), None) => other.clone(),
+            (None, Some(false)) => self.clone(),
+            (None, None) => {
+                /*
+                   (1 - 2 a) (1 - 2 b) = 1 - 2 c
+                1 - 2 (a + b) + 4 a b = 1 - 2 c
+                - 2 (a + b) + 4 a b = - 2 c
+                (a + b) - 2 a b = c
+                2 a b = a + b - c
+                 */
+
+                let self_clone = self.clone();
+                let other_clone = other.clone();
+                let res: Boolean<F> = state.compute(TypeCreation::Unsafe, loc!(), move |env| {
+                    let b1: bool = self_clone.read(env);
+                    let b2: bool = other_clone.read(env);
+
+                    /*
+                    let%bind res =
+                      exists typ_unchecked
+                        ~compute:
+                          As_prover.(
+                            map2 ~f:Bool.( <> ) (read typ_unchecked b1)
+                              (read typ_unchecked b2))
+                    in
+                     */
+
+                    todo!()
+                });
+
+                let x = &self.0 + &self.0;
+                let y = &other.0;
+                let z = &self.0 + &other.0 - &res.0;
+
+                // TODO: annotation?
+                state.assert_r1cs(None, x, y.clone(), z);
```

Comment:
Nit: unnecessary clone

Reply:
necessary because we have a reference here (btw clippy warns us when clones are unnecessary :o)